packages/dd-trace/src/appsec/api_security_sampler.js

```diff
@@ -1,61 +1,87 @@
 'use strict'
 
+const LRUCache = require('lru-cache')
+const web = require('../plugins/util/web')
 const log = require('../log')
+const { AUTO_REJECT, USER_REJECT } = require('../../../../ext/priority')
 
-let enabled
-let requestSampling
+const MAX_SIZE = 4096
+const DEFAULT_DELAY = 30 // 30s
 
-const sampledRequests = new WeakSet()
+let enabled
+let sampledRequests
 
 function configure ({ apiSecurity }) {
   enabled = apiSecurity.enabled
-  setRequestSampling(apiSecurity.requestSampling)
+  const delay = apiSecurity.sampleDelay || DEFAULT_DELAY
+  sampledRequests = new LRUCache({ max: MAX_SIZE, ttl: delay * 1000 })
 }
 
 function disable () {
   enabled = false
+  sampledRequests?.clear()
 }
 
-function setRequestSampling (sampling) {
-  requestSampling = parseRequestSampling(sampling)
-}
+function sampleRequest (req, res, force = false) {
+  if (!enabled) return false
+  if (this.isSampled(req, res)) return false
 
-function parseRequestSampling (requestSampling) {
-  let parsed = parseFloat(requestSampling)
+  const rootSpan = web.root(req)
+  if (!rootSpan) return false
 
-  if (isNaN(parsed)) {
-    log.warn(`Incorrect API Security request sampling value: ${requestSampling}`)
+  const priority = getSpanPriority(rootSpan)
 
-    parsed = 0
-  } else {
-    parsed = Math.min(1, Math.max(0, parsed))
+  if (priority === AUTO_REJECT || priority === USER_REJECT) {
+    return false
   }
 
-  return parsed
-}
-
-function sampleRequest (req) {
-  if (!enabled || !requestSampling) {
+  if (!priority && !rootSpan._prioritySampler?.isSampled(rootSpan)) {
```

Comment:
now we are calling rootSpan._prioritySampler?.sample(rootSpan) so this block must be removed
```javascript
// remove it
if (!priority && !rootSpan._prioritySampler?.isSampled(rootSpan)) {
  return false
}
```
(github UI doesn't allow to do a proper suggestion because it contains deleted lines)

```suggestion

```